vortex-zigzag/src/compress.rs

```diff
@@ -30,87 +27,119 @@ impl EncodingCompression for ZigZagEncoding {
     }
 
     fn compress(
         &self,
-        array: &dyn Array,
-        like: Option<&dyn Array>,
+        array: &Array,
+        like: Option<&Array>,
         ctx: CompressCtx,
-    ) -> VortexResult<ArrayRef> {
-        let zigzag_like = like.map(|like_arr| like_arr.as_zigzag());
-        let encoded = match ArrayKind::from(array) {
-            ArrayKind::Primitive(p) => zigzag_encode(p),
-            _ => unreachable!("This array kind should have been filtered out"),
-        }
-        .unwrap();
+    ) -> VortexResult<OwnedArray> {
+        let zigzag_like = like.map(|like_arr| like_arr.as_array_ref());
+        let encoded = if array.encoding().id() == Primitive::ID {
+            zigzag_encode(&array.as_primitive())?
+        } else {
+            println!("{}", array.encoding().id());
+            unreachable!("This array kind should have been filtered out")
+        };
 
-        Ok(
-            ZigZagArray::new(ctx.compress(encoded.encoded(), zigzag_like.map(|z| z.encoded()))?)
-                .into_array(),
-        )
+        Ok(OwnedZigZagArray::new(ctx.compress(&encoded.encoded(), zigzag_like)?).into_array())
     }
 }
 
-pub fn zigzag_encode(parray: &PrimitiveArray) -> VortexResult<ZigZagArray> {
+pub fn zigzag_encode(parray: &PrimitiveArray<'_>) -> VortexResult<OwnedZigZagArray> {
     let encoded = match parray.ptype() {
-        PType::I8 => zigzag_encode_primitive::<i8>(parray.buffer().typed_data(), parray.validity()),
+        PType::I8 => {
+            zigzag_encode_primitive::<i8>(parray.buffer().typed_data(), Some(parray.validity()))
+        }
         PType::I16 => {
-            zigzag_encode_primitive::<i16>(parray.buffer().typed_data(), parray.validity())
+            zigzag_encode_primitive::<i16>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         PType::I32 => {
-            zigzag_encode_primitive::<i32>(parray.buffer().typed_data(), parray.validity())
+            zigzag_encode_primitive::<i32>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         PType::I64 => {
-            zigzag_encode_primitive::<i64>(parray.buffer().typed_data(), parray.validity())
+            zigzag_encode_primitive::<i64>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         _ => panic!("Unsupported ptype {}", parray.ptype()),
     };
-    ZigZagArray::try_new(encoded.into_array())
+    OwnedZigZagArray::try_new(encoded.into_array())
 }
 
-fn zigzag_encode_primitive<T: ZigZag + NativePType>(
-    values: &[T],
-    validity: Option<ValidityView>,
-) -> PrimitiveArray
+fn zigzag_encode_primitive<'a, T: ExternalZigZag + NativePType>(
+    values: &'a [T],
+    validity: Option<Validity<'a>>,
+) -> PrimitiveArray<'a>
 where
-    <T as ZigZag>::UInt: NativePType,
+    <T as ExternalZigZag>::UInt: NativePType,
 {
     let mut encoded = AlignedVec::with_capacity_in(values.len(), ALIGNED_ALLOCATOR);
     encoded.extend(values.iter().map(|v| T::encode(*v)));
```

Comment:
There's no point using AlignedVec if you then create a copy with to_vec below. 

For now, just get rid of the AlignedVec stuff, I think we're going to defer handling alignment for now.